### PR TITLE
check for conflicting revision and resolve with the most recent version

### LINKED ISSUE #43: always use remote docs when replicating integrated data
When integrated data is updated out of sequence (e.g. by an import) remotely local replication will frequently choose the local doc, as the imported doc will have a lower rev owing to its being created from scratch. 

This leads to changes by etl (e.g. importing missed levels) being ignored by local databases.

- [ ] replication of integrated data should always accept remote changes regardless of difference in revision number

### PR DESCRIPTION
**Pull request by freefony**
closes #43

### CODE REVIEW

File: .idea/angular-nav-data.iml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tlvince) [inline .idea/angular-nav-data.iml]: Should you commit these IDE things?

File: src/locations/locations.service.js [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import { replication as replicationConfig } from '../config.json'
+import { default as utilService } from '../utils/utils.service'
```
Comment (tlvince) [inline src/locations/locations.service.js]: As `checkAndResolveConflicts` is a pure function and doesn't rely on any class members in `utilService`, I'd move it out as a normal `export` or into a non-Angular wrapped file.

File: src/locations/locations.service.js [hunk 3]
```diff
@@ -62,6 +63,8 @@ class LocationsService {
       this.replicationFrom
         .on('paused', onReplicationPaused)
     }
+
+    this.localDB.changes({conflicts: true, onChange: utilService.checkAndResolveConflicts.bind(null, this.localDB)})
```
Comment (tlvince) [inline src/locations/locations.service.js]: Did you try without `bind`?

```diff
- this.localDB.changes({conflicts: true, onChange: utilService.checkAndResolveConflicts.bind(null, this.localDB)})
+ this.localDB.changes({conflicts: true, onChange: change => utilService.checkAndResolveConflicts(change, this.localDB)})
```

File: src/utils/utils.service.js [hunk 4]
```diff
@@ -59,6 +59,47 @@ class UtilsService {
...
+  checkAndResolveConflicts (changedDoc, pouchdb) {
+    let preferredRevision = {}
+    if (!changedDoc._conflicts) {
+      return
+    }
+
+    if (!changedDoc.updatedAt) {
+      preferredRevision = changedDoc
+      changedDoc._conflicts.map((conflictingRev) => {
+        if (conflictingRev.updatedAt) {
```
Comment (tlvince) [inline src/utils/utils.service.js]: Shouldn't `updateAt` be checked to see if it's newer than the currently chosen `preferredRevision`?

File: src/utils/utils.service.js [hunk 5]
```diff
@@ -59,6 +59,47 @@ class UtilsService {
...
+      let serializedRevisions = this.serialiseDocWithConflictsByProp(changedDoc, 'updatedAt')
+      preferredRevision = serializedRevisions.pop()
+      serializedRevisions.forEach(revision => {
+        pouchdb.remove(changedDoc._id, revision)
+      })
+    }
+    pouchdb.put(preferredRevision) // do we still need to PUT after plucking conflicts?
+  }
+
+  serialiseDocWithConflicts (Doc, prop) {
```
Comment (tlvince) [inline src/utils/utils.service.js]: Why uppercase `Doc`? It's not a constructor.

File: src/utils/utils.service.js [hunk 6]
```diff
@@ -59,6 +59,47 @@ class UtilsService {
...
+    let mainDoc = {}
+    let serialisedDocs = []
+    Object.keys(Doc).map((prop) => {
```
Comment (tlvince) [inline src/utils/utils.service.js]: Why iterate over all keys if you're only looking for `_conflicts`? Can't you do `if (Doc._conflicts) mainDoc._conflicts = Doc._conflicts`?

File: src/utils/utils.service.js [hunk 7]
```diff
@@ -59,6 +59,47 @@ class UtilsService {
...
+      preferredRevision = changedDoc
+      changedDoc._conflicts.map((conflictingRev) => {
+        if (conflictingRev.updatedAt) {
+          preferredRevision = conflictingRev
+        } else {
+          pouchdb.remove(changedDoc._id, conflictingRev._rev)
+        }
+      })
+    } else {
+      let serializedRevisions = this.serialiseDocWithConflictsByProp(changedDoc, 'updatedAt')
```
Comment (tlvince) [inline src/utils/utils.service.js]: So this sorts conflicts by `updateAt`?

File: src/utils/utils.service.js [hunk 8]
```diff
@@ -59,6 +59,47 @@ class UtilsService {
...
+    let serialisedDocs = []
+    Object.keys(Doc).map((prop) => {
+      if (prop === '_conflicts') {
+        mainDoc[prop] = Doc[prop]
+      }
+    })
+    serialisedDocs = [mainDoc]
+      .concat(Doc._conflicts)
+      .sort((a, b) => {
+        return a[prop] > b[prop]
```
Comment (tlvince) [inline src/utils/utils.service.js]: This will always return `false` if `prop` (`updatedAt`) is `undefined`

**Review by tlvince — CHANGES_REQUESTED** [review]
The logic seems straightforward enough, but I think we need some integration tests here to be confident it works as intended.

**Comment by tlvince** [comment]
@freefony I took the liberty of writing a (failing) integration test. Please could you check the assertion and confirm whether my understand aligns with yours?

Also note I fixed a bug in the locations service whereby you need to use the injected version of the utils service (4886f2d). If you agree with that change, you'll also need to apply it to the other services.

(You can ignore the `chore` scopes above. They aren't directly related to this PR, but were necessary in order to run the integration test).

File: src/utils/utils.service.js (hunk 4 above)
Comment (freefony) [reply src/utils/utils.service.js]: the `if` checks that the `changedDoc` does not have `updatedAt` property so at this point `preferredRevision` does not have an `updatedAt` property so cannot be newer

File: src/utils/utils.service.js (hunk 7 above)
Comment (freefony) [reply src/utils/utils.service.js]: Yes. also adds the main doc (couchdb's "winning" rev) so all are in an array and can be sorted

**Comment by tlvince** [comment]
@freefony I've added restart replication support to locations service and made sure the local rev is at `2-`. However, so far it doesn't even look like the `onChange` handler is called.

File: src/utils/utils.service.js (hunk 6 above)
Comment (freefony) [reply src/utils/utils.service.js]: Was trying to copy the doc without the `_conflicts` array property i fixed this in an update commit

**Comment by tlvince** [comment]
@freefony I fixed the change handler invocation in 3053a8f. Note angular-pouchdb's API is slightly different here.

**Comment by tlvince** [comment]
The test still fails on `Error: undefined is not a constructor (evaluating 'this.serialiseDocWithConfl
ictsByProp(changedDoc, 'updatedAt')')`

**Comment by tlvince** [comment]
That's fixed, but this is still not working as we'd expect as the `_conflicts` property is indeed a list of revs, not a list of docs. Note we're using PouchDB 5.4.x, not 6.x which the PouchDB docs are based on (might've been a source of confusion).

**Comment by freefony** [comment]
Ok. i ll do `db.get` on each doc then

**Comment by tlvince** [comment]
OK, NB, you don't need to fetch all rev docs by hand; you can use `open_revs: [<rev>] | 'all'` instead, e.g.:

```js
if (doc._conflicts.length) {
  db.get(doc._id, { open_revs: doc._conflicts })
  //=> [{}, {}, …]
}
```

Or indeed if there's only one conflict, use `rev: <rev>`.

File: test/locations.service.integration.js [hunk 9]
```diff
@@ -17,7 +17,7 @@ function shouldNotBeCalled (rejection) {
 }
 
 describe('locationsService', function () {
-  var url = window.__env__.COUCHDB_URL || 'http://localhost:5984'
```
Comment (tlvince) [inline test/locations.service.integration.js]: Was this causing any issues? I added it for ease of testing locally (can just run `npm t`)

**Review by tlvince — CHANGES_REQUESTED** [review]
~~The test is failing for me with `Expected '2-' to be '3-'.`.~~

Actually, hold on, CouchDB wasn't started.

File: src/utils/utils.service.js [hunk 10]
```diff
@@ -59,6 +59,51 @@ class UtilsService {
...
+    if (!changedDoc._conflicts) {
+      return
+    }
+
+    pouchdb.get(changedDoc._id, {'open_revs': changedDoc._conflicts})
+      .then(conflictingRevObjs => {
+        let serializedRevisions = this.serialiseDocWithConflictsByProp(changedDoc, conflictingRevObjs, 'updatedAt')
+        preferredRevision = serializedRevisions.shift()
+        serializedRevisions.forEach(revision => {
+          pouchdb.remove(changedDoc._id, revision._rev)
```
Comment (tlvince) [inline src/utils/utils.service.js]: Could you do this a `bulkDocs` perhaps?

File: src/utils/utils.service.js [hunk 11]
```diff
@@ -59,6 +59,51 @@ class UtilsService {
...
+            .catch(err => console.error(err))
+        })
+        preferredRevision._conflicts = []
+        pouchdb.put(preferredRevision)
+          .catch(err => {
```
Comment (tlvince) [inline src/utils/utils.service.js]: Probably we should bubble these up to the caller so it can respond accordingly.

**Review by tlvince — CHANGES_REQUESTED** [review]
The test is intermittently failing as `callOnReplicationComplete` occasionally resolves before `pouchdb.put(preferredRevision)`.

One way to solve this would be to add callback registration on the change handler similar to how's done for replication. That'd also be a good place to bubble any errors to.

**Comment by tlvince** [comment]
@freefony I'm going to have look at this intermittent test.

**Review by tlvince — APPROVED** [review]

**Comment by tlvince** [comment]
@freefony this lgtm now. Please have a run through a feel free to merge if it lgty 👍

**Pull request merged**